File: app/src/main/java/com/master/kit/utils/FragmentSwitchManager.java
package com.master.kit.utils;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;

import static android.R.attr.id;


/**
 * Created by master on 2017/9/27 0027.
 */

public class FragmentSwitchManager {
    FragmentManager manager;
    int[] tags;
    int containerViewId;
    FragmentSwitchListener listener;
    /**
     *
     * @param manager
     * @param containerViewId
     * @param tagId   fragment的tag;建议以按钮的id作为tagId
     */
    public FragmentSwitchManager(FragmentSwitchListener listener, FragmentManager manager, int containerViewId, int... tagId) {
        this.manager = manager;
        tags = tagId;
        this.containerViewId = containerViewId;
        this.listener = listener;
    }

    public void switchFragment(int tagId) {
        String tag = String.valueOf(tagId);
        FragmentTransaction fragmentTransaction = manager.beginTransaction();
        hideAll(fragmentTransaction);
        Fragment fragment = manager.findFragmentByTag(tag);
        if (null == fragment) {
            fragment = listener.createFragment(tagId);
            fragmentTransaction.add(containerViewId, fragment, tag);
        }
        listener.onSwitchFragment(tagId,fragment);
        fragmentTransaction.show(fragment).commitAllowingStateLoss();
    }

    private void hideAll(FragmentTransaction fragmentTransaction) {
        for (int i = 0; i < tags.length; i++) {
            Fragment fragment;
            fragment = manager.findFragmentByTag(String.valueOf(tags[i]));
            if (fragment != null) {
                fragmentTransaction.hide(fragment);
            }
        }
    }
    public interface FragmentSwitchListener {
        Fragment createFragment(int id);
        void onSwitchFragment(int id,Fragment fragment);
    }
}

File: service/src/main/java/tf/oof/com/service/engine/FragmentSwitchManager.kt
package tf.oof.com.service.engine

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentTransaction


/**
 * Created by master on 2017/9/27 0027.
 */
/**
 *
 * @param manager
 * @param containerViewId
 * @param tagId   fragment的tag;建议以按钮的id作为tagId
 */
class FragmentSwitchManager(
        private var manager: FragmentManager,
        private var containerViewId: Int,
        vararg tagId: Int,
        private val onDoSwitch: ((Int, Fragment) -> Unit)?,
        private val createFragment: (Int) -> Fragment) {
    private var tags: IntArray = tagId

    fun switchFragment(tagId: Int) {
        val fragmentTransaction = manager.beginTransaction()
        hideAll(fragmentTransaction)
        var fragment: Fragment = findFragment(fragmentTransaction, tagId)
        onDoSwitch?.invoke(tagId, fragment)
        fragmentTransaction.show(fragment).commitAllowingStateLoss()
    }

    private fun hideAll(fragmentTransaction: FragmentTransaction) {
        tags.map { manager.findFragmentByTag(it.toString()) }
                .filter { it != null }
                .forEach { fragmentTransaction.hide(it) }
    }

    private fun findFragment(fragmentTransaction: FragmentTransaction, tagId: Int): Fragment {
        var fragment: Fragment? = manager.findFragmentByTag(tagId.toString())
        if (null == fragment) {
            fragment = createFragment(tagId)
            fragmentTransaction.add(containerViewId, fragment, tagId.toString())
        }
        return fragment
    }
}
